File: 1140-stone-game-ii/1140-stone-game-ii.py
class Solution:
    def stoneGameII(self, piles):
        n = len(piles)

        suffix = [0] * (n + 1)

        for i in range(n - 1, -1, -1):
            suffix[i] = suffix[i + 1] + piles[i]

        dp = [[-1] * (n + 1) for _ in range(n)]

        def solve(i, M):
            if i >= n:
                return 0

            if i + 2 * M >= n:
                return suffix[i]

            if dp[i][M] != -1:
                return dp[i][M]

            best = 0

            for x in range(1, 2 * M + 1):
                opponent = solve(i + x, max(M, x))

                best = max(
                    best,
                    suffix[i] - opponent
                )

            dp[i][M] = best
            return best

        return solve(0, 1)

File: 1140-stone-game-ii/1140-stone-game-ii.java
class Solution {
    int n;
    int[] suffix;
    int[][] dp;

    private int solve(int i, int M) {
        if (i >= n)
            return 0;

        if (i + 2 * M >= n)
            return suffix[i];

        if (dp[i][M] != -1)
            return dp[i][M];

        int best = 0;

        for (int x = 1; x <= 2 * M; ++x) {
            int opponent = solve(i + x, Math.max(M, x));

            best = Math.max(
                best,
                suffix[i] - opponent
            );
        }

        return dp[i][M] = best;
    }

    public int stoneGameII(int[] piles) {
        n = piles.length;

        suffix = new int[n + 1];

        for (int i = n - 1; i >= 0; --i) {
            suffix[i] = suffix[i + 1] + piles[i];
        }

        dp = new int[n][n + 1];

        for (int i = 0; i < n; ++i) {
            java.util.Arrays.fill(dp[i], -1);
        }

        return solve(0, 1);
    }
}